### DISCUSSION TITLE
how can i do an api test for theia-blueprint apps?

### DISCUSSION THREAD

**Discussion by MrBen99** [eclipse-theia/theia#13511] [Q&A] [answered]
we use  theia-blueprint to build our desktop apps. i know theia will start a websocket service, but i don't know how to create a connection, send and receive the message.
![image](https://github.com/eclipse-theia/theia/assets/94780689/2e043d0d-11c2-48a4-b1f5-298b3b342d85)
![image](https://github.com/eclipse-theia/theia/assets/94780689/6b913ab4-1f5c-40d1-8bd5-8cdfe0fb9c28)

**Comment by JonasHelming** [comment]
Hi, could you describe more in detail what you want to achieve, please?

**Reply by MrBen99** [reply]
Hi Jonas，
Thanks for your reply.
i'm a tester from my project. 
what i get is an installable file, not the source project. when i  install the app, it's just like theia-blueprint. 
when i start the app, and from the developer tools, i can get the process  ip and port, 
![image](https://github.com/eclipse-theia/theia/assets/94780689/83afc46e-18a4-4cff-aa9a-55362169e27c)

but i don't know how to connect it by ip and port , and send and receive the message.
i try to use postman to connect it to do the api test,but it failed.
![image](https://github.com/eclipse-theia/theia/assets/94780689/39b122e8-4a50-4793-b9cc-8933a2713f01)

do you have any ideal for this situation? or is there any way i can connect the theia backend by ip and port for the api test?

**Reply by msujew** [reply, marked as answer]
Theia doesn't employ a REST api but an (rather undocumented) custom RPC mechanism employing a binary message passing protocol. Using it to test the server part of your app will be incredibly complicated. In fact, incoming requests to the electron backend will be rejected if they don't originate from the electron window.

In order to test your app, you should use something like the Theia playright package. See [here](https://github.com/eclipse-theia/theia-playwright-template) for more info.

If you want to test the server on its own, my recommendation would be integration tests with jest or mocha, rather than end-to-end tests.

**Reply by MrBen99** [reply]
thanks very much
